File: py/567.py
import string


class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        ms1 = {}
        ms2 = {}

        for c in string.ascii_lowercase:
            ms1[c] = 0
            ms2[c] = 0

        for c in s1:
            ms1[c] += 1

        matches = 0
        for c in string.ascii_lowercase:
            if ms1[c] == ms2[c]:
                matches += 1

        l = 0
        for r in range(len(s2)):
            c = s2[r]
            ms2[c] += 1
            if ms2[c] == ms1[c]:
                matches += 1
            if r < len(s1) - 1:
                continue

            if matches == 26:
                return True

            c = s2[l]

            if ms1[c] == ms2[c]:
                matches -= 1
            ms2[c] -= 1
            l += 1

        return False

File: go/567.go
package main

func checkInclusion(s1 string, s2 string) bool {
	ms1 := [26]int{}
	ms2 := [26]int{}

	for _, c := range s1 {
		ms1[c-'a']++
	}
	l := 0
	for r := range s2 {
		ms2[s2[r]-'a']++
		if r < len(s1)-1 {
			continue
		}

		if is_equal(ms1, ms2) {
			return true
		}
		ms2[s2[l]-'a']--
		l++
	}

	return false
}

func is_equal(ms1, ms2 [26]int) bool {
	for i := range ms1 {
		if ms1[i] != ms2[i] {
			return false
		}
	}

	return true
}
